File: leetcode/Jump_Game.java
import java.util.*;

public class Solution {
    public boolean canJump(int[] A) {
        int p = 0;
        for (int i = 0; i < A.length; ++i)
            if (i <= p) {
                p = Math.max(p, i + A[i]);
                if (p >= A.length - 1) return true;
            }
        return false;
    }
}

File: leetcode/Jump_Game.cpp
class Solution {
public:
    bool canJump(int A[], int n) {
        if (n == 0)
            return false;
        if (n == 1)
            return true;
        int rightmost = 0;
        for (int i = 0; i < n - 1; ++i)
            if (i <= rightmost) {  // reachable
                rightmost = max(rightmost, i + A[i]);
                if (rightmost >= n - 1)
                    return true;
            } else {
                return false;
            }
        return false;
    }
};
